The job **`i686-mingw`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10949555656/job/30403060694) [(plain)](https://github.com/rust-lang/rust/commit/72d4f3c0b6c866c152a17dfa09232f3ee10d1e32/checks/30403060694/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[command]C:\Windows\system32\cmd.exe /D /S /C C:\a\_temp\setup-msys2\msys2.cmd -c "'paccache' '-r' '-f' '-u' '-k0'"
==> no candidate packages found for pruning
[command]C:\Windows\system32\cmd.exe /D /S /C C:\a\_temp\setup-msys2\msys2.cmd -c "'paccache' '-r' '-f' '-k1'"
==> no candidate packages found for pruning
[command]"C:\Program Files\Git\usr\bin\tar.exe" --posix -cf cache.tzst --exclude cache.tzst -P -C C:/a/rust/rust --files-from manifest.txt --force-local --use-compress-program "zstd -T0"
Cache saved successfully
Cache saved as ID 229 using key msys2-pkgs-upd:false-conf:e8555627-files:305123cd29efffaad88eb14e88ff433b93ed4169
[command]C:\Windows\system32\cmd.exe /D /S /C C:\a\_temp\setup-msys2\msys2.cmd -c "'pacman' '--noconfirm' '-Scc'"

---
  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 7110869d Merge 72d4f3c0b6c866c152a17dfa09232f3ee10d1e32 into 506f22b4663f3e756e1e6a4f66c6309fdc00819c
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format='%H'
'7110869d01720b35f9b904e19c9a83ca6c29e0c1'
##[group]Run echo "[CI_PR_NUMBER=$num]"
echo "[CI_PR_NUMBER=$num]"
---
   Compiling miniz_oxide v0.7.4
   Compiling addr2line v0.22.0
error: linking with `i686-w64-mingw32-gcc` failed: exit code: 1
  |
  = note: "i686-w64-mingw32-gcc" "-Wl,C:\\a\\_temp\\msys64\\tmp\\rustcZx43Yw\\list.def" "-fno-use-linker-plugin" "-Wl,--dynamicbase" "-Wl,--disable-auto-image-base" "-Wl,--large-address-aware" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\rsbegin.o" "C:\\a\\_temp\\msys64\\tmp\\rustcZx43Yw\\symbols.o" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\std-ef332fa5d4567ec2.4qe06mw0zt8hm9snjzeh095uu.rcgu.rmeta" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\std-ef332fa5d4567ec2.9z7mvged9kga5csmkp3bny003.rcgu.o" "-Wl,-Bdynamic" "-lkernel32" "-lkernel32" "-Wl,-Bstatic" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libpanic_unwind-27702dcdf49b38e3.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libobject-39b113a4ef159b13.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libmemchr-76a615b0b978ca93.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libaddr2line-4d9cdfb0d790e1d1.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libgimli-6d42266be6eee9e3.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libwindows_targets-e65bfd93301c232f.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\librustc_demangle-0d7bc50d3e550d3a.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libstd_detect-2c6653e355aa5f3d.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libhashbrown-3870830ff3b2e5ef.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\librustc_std_workspace_alloc-a03bbbc17988b2d8.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libminiz_oxide-9484d2ad05333da0.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libadler-41cdc59204f6f9da.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libunwind-058a1f651c3be62c.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libcfg_if-a6c4e33c45cb0994.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\liblibc-c718a0c128385429.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\liballoc-2eb7b655f60e9152.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\librustc_std_workspace_core-4af3e43379c9cf85.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libcore-38080d2d1e9e8271.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libcompiler_builtins-fce873f7fbdc0cf5.rlib" "-Wl,-Bdynamic" "-ladvapi32" "-lntdll" "-luserenv" "-lws2_32" "-ldbghelp" "C:\\a\\_temp\\msys64\\tmp\\rustcZx43Yw\\bcryptprimitives.dll_imports.lib" "C:\\a\\_temp\\msys64\\tmp\\rustcZx43Yw\\api-ms-win-core-synch-l1-2-0.dll_imports.lib" "-lgcc_s" "-lmsvcrt" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-lmingwex" "-luser32" "-lkernel32" "-Wl,--nxcompat" "-L" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\build\\compiler_builtins-9168138e418349cb\\out" "-o" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\std-ef332fa5d4567ec2.dll" "-shared" "-Wl,--out-implib=C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1-std\\i686-pc-windows-gnu\\release\\deps\\libstd-ef332fa5d4567ec2.dll.a" "-Wl,-O1" "-nodefaultlibs" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage1\\lib\\rustlib\\i686-pc-windows-gnu\\lib\\rsend.o"
  = note: C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x2726): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x3a67): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x3af9): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x4910): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x51e5): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x2ee8a): more undefined references to `_imp_WakeByAddressSingle@4' follow␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x2f58f): undefined reference to `_imp_WaitOnAddress@16'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x2f72d): undefined reference to `_imp_WaitOnAddress@16'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x3091e): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x3788f): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x379c8): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x37a88): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x37b53): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x37c05): more undefined references to `_imp_WakeByAddressSingle@4' follow␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x4427a): undefined reference to `_imp_WakeByAddressAll@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x442e5): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x442f9): undefined reference to `_imp_WaitOnAddress@16'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x44369): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x444ac): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x444cc): undefined reference to `_imp_WakeByAddressAll@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x44873): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x44b04): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x52f3e): undefined reference to `_imp_ProcessPrng@8'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x588f8): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5f77a): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5f78c): undefined reference to `_imp_WaitOnAddress@16'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5f7e3): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5f868): undefined reference to `_imp_WaitOnAddress@16'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5f8f8): undefined reference to `_imp_WaitOnAddress@16'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5f958): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5f975): undefined reference to `_imp_WakeByAddressAll@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5f9a5): undefined reference to `_imp_WaitOnAddress@16'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5f9f0): undefined reference to `_imp_WaitOnAddress@16'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5fa9c): undefined reference to `_imp_WaitOnAddress@16'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5fb46): undefined reference to `_imp_WaitOnAddress@16'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5fbcd): undefined reference to `_imp_WakeByAddressAll@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5fbe9): undefined reference to `_imp_WakeByAddressAll@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5fc44): undefined reference to `_imp_WaitOnAddress@16'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5fdea): undefined reference to `_imp_WaitOnAddress@16'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5fe8e): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5fecc): undefined reference to `_imp_WakeByAddressSingle@4'␍
          C:/a/rust/rust/mingw32/bin/../lib/gcc/i686-w64-mingw32/14.1.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\a\rust\rust\build\i686-pc-windows-gnu\stage1-std\i686-pc-windows-gnu\release\deps\std-ef332fa5d4567ec2.std.b03eed99e6cdbeea-cgu.0.rcgu.o:std.b03eed99e6cdbeea-cgu.0:(.text+0x5fee0): undefined reference to `_imp_WakeByAddressAll@4'␍
          collect2.exe: error: ld returned 1 exit status
          
  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#rustc-link-lib)
error: could not compile `std` (lib) due to 1 previous error
Build completed unsuccessfully in 0:25:04
make: *** [Makefile:115: ci-mingw-bootstrap] Error 1
  local time: Thu, Sep 19, 2024 10:43:45 PM
```

</details>

--- a/compiler/rustc_codegen_gcc/src/archive.rs
+++ b/compiler/rustc_codegen_gcc/src/archive.rs
@@ -2,6 +2,7 @@ use std::path::Path;
 
 use rustc_codegen_ssa::back::archive::{
     ArArchiveBuilder, ArchiveBuilder, ArchiveBuilderBuilder, DEFAULT_OBJECT_READER,
+    ImportLibraryItem,
 };
 use rustc_session::Session;
 
@@ -16,7 +17,7 @@ impl ArchiveBuilderBuilder for ArArchiveBuilderBuilder {
         &self,
         _sess: &Session,
         _lib_name: &str,
-        _import_name_and_ordinal_vector: Vec<(String, Option<u16>)>,
+        _items: Vec<ImportLibraryItem>,
         _output_path: &Path,
     ) {
         unimplemented!("creating dll imports is not yet supported");

--- a/compiler/rustc_codegen_llvm/src/callee.rs
+++ b/compiler/rustc_codegen_llvm/src/callee.rs
@@ -49,6 +49,22 @@ pub(crate) fn get_fn<'ll, 'tcx>(cx: &CodegenCx<'ll, 'tcx>, instance: Instance<'t
         let llfn = if tcx.sess.target.arch == "x86"
             && let Some(dllimport) = crate::common::get_dllimport(tcx, instance_def_id, sym)
         {
+            // When calling functions in generated import libraries, MSVC needs
+            // the fully decorated name (as would have been in the declaring
+            // object file), but MinGW wants the name as exported (as would be
+            // in the def file) which may be missing decorations.
+            let mingw_gnu_toolchain = common::is_mingw_gnu_toolchain(&tcx.sess.target);
+            let llfn = cx.declare_fn(
+                &common::i686_decorated_name(
+                    dllimport,
+                    mingw_gnu_toolchain,
+                    true,
+                    !mingw_gnu_toolchain,
+                ),
+                fn_abi,
+                Some(instance),
+            );
+
             // Fix for https://github.com/rust-lang/rust/issues/104453
             // On x86 Windows, LLVM uses 'L' as the prefix for any private
             // global symbols, so when we create an undecorated function symbol
@@ -60,15 +76,6 @@ pub(crate) fn get_fn<'ll, 'tcx>(cx: &CodegenCx<'ll, 'tcx>, instance: Instance<'t
             // LLVM will prefix the name with `__imp_`. Ideally, we'd like the
             // existing logic below to set the Storage Class, but it has an
             // exemption for MinGW for backwards compatibility.
-            let llfn = cx.declare_fn(
-                &common::i686_decorated_name(
-                    dllimport,
-                    common::is_mingw_gnu_toolchain(&tcx.sess.target),
-                    true,
-                ),
-                fn_abi,
-                Some(instance),
-            );
             unsafe {
                 llvm::LLVMSetDLLStorageClass(llfn, llvm::DLLStorageClass::DllImport);
             }

--- a/compiler/rustc_codegen_llvm/src/consts.rs
+++ b/compiler/rustc_codegen_llvm/src/consts.rs
@@ -196,16 +196,10 @@ fn check_and_apply_linkage<'ll, 'tcx>(
             g2
         }
     } else if cx.tcx.sess.target.arch == "x86"
+        && common::is_mingw_gnu_toolchain(&cx.tcx.sess.target)
         && let Some(dllimport) = crate::common::get_dllimport(cx.tcx, def_id, sym)
     {
-        cx.declare_global(
-            &common::i686_decorated_name(
-                dllimport,
-                common::is_mingw_gnu_toolchain(&cx.tcx.sess.target),
-                true,
-            ),
-            llty,
-        )
+        cx.declare_global(&common::i686_decorated_name(dllimport, true, true, false), llty)
     } else {
         // Generate an external declaration.
         // FIXME(nagisa): investigate whether it can be changed into define_global

--- a/compiler/rustc_codegen_ssa/src/back/archive.rs
+++ b/compiler/rustc_codegen_ssa/src/back/archive.rs
@@ -26,6 +26,35 @@ use crate::errors::{
     DlltoolFailImportLibrary, ErrorCallingDllTool, ErrorCreatingImportLibrary, ErrorWritingDEFFile,
 };
 
+/// An item to be included in an import library.
+/// This is a slimmed down version of `COFFShortExport` from `ar-archive-writer`.
+pub struct ImportLibraryItem {
+    /// The name to be exported.
+    pub name: String,
+    /// The ordinal to be exported, if any.
+    pub ordinal: Option<u16>,
+    /// The original, decorated name if `name` is not decorated.
+    pub symbol_name: Option<String>,
+    /// True if this is a data export, false if it is a function export.
+    pub is_data: bool,
+}
+
+impl From<ImportLibraryItem> for COFFShortExport {
+    fn from(item: ImportLibraryItem) -> Self {
+        COFFShortExport {
+            name: item.name,
+            ext_name: None,
+            symbol_name: item.symbol_name,
+            alias_target: None,
+            ordinal: item.ordinal.unwrap_or(0),
+            noname: item.ordinal.is_some(),
+            data: item.is_data,
+            private: false,
+            constant: false,
+        }
+    }
+}
+
 pub trait ArchiveBuilderBuilder {
     fn new_archive_builder<'a>(&self, sess: &'a Session) -> Box<dyn ArchiveBuilder + 'a>;
 
@@ -38,7 +67,7 @@ pub trait ArchiveBuilderBuilder {
         &self,
         sess: &Session,
         lib_name: &str,
-        import_name_and_ordinal_vector: Vec<(String, Option<u16>)>,
+        items: Vec<ImportLibraryItem>,
         output_path: &Path,
     ) {
         if common::is_mingw_gnu_toolchain(&sess.target) {
@@ -47,21 +76,16 @@ pub trait ArchiveBuilderBuilder {
             // that loaded but crashed with an AV upon calling one of the imported
             // functions. Therefore, use binutils to create the import library instead,
             // by writing a .DEF file to the temp dir and calling binutils's dlltool.
-            create_mingw_dll_import_lib(
-                sess,
-                lib_name,
-                import_name_and_ordinal_vector,
-                output_path,
-            );
+            create_mingw_dll_import_lib(sess, lib_name, items, output_path);
         } else {
             trace!("creating import library");
             trace!("  dll_name {:#?}", lib_name);
             trace!("  output_path {}", output_path.display());
             trace!(
                 "  import names: {}",
-                import_name_and_ordinal_vector
+                items
                     .iter()
-                    .map(|(name, _ordinal)| name.clone())
+                    .map(|ImportLibraryItem { name, .. }| name.clone())
                     .collect::<Vec<_>>()
                     .join(", "),
             );
@@ -79,20 +103,7 @@ pub trait ArchiveBuilderBuilder {
                     .emit_fatal(ErrorCreatingImportLibrary { lib_name, error: error.to_string() }),
             };
 
-            let exports = import_name_and_ordinal_vector
-                .iter()
-                .map(|(name, ordinal)| COFFShortExport {
-                    name: name.to_string(),
-                    ext_name: None,
-                    symbol_name: None,
-                    alias_target: None,
-                    ordinal: ordinal.unwrap_or(0),
-                    noname: ordinal.is_some(),
-                    data: false,
-                    private: false,
-                    constant: false,
-                })
-                .collect::<Vec<_>>();
+            let exports = items.into_iter().map(Into::into).collect::<Vec<_>>();
             let machine = match &*sess.target.arch {
                 "x86_64" => MachineTypes::AMD64,
                 "x86" => MachineTypes::I386,
@@ -160,16 +171,16 @@ pub trait ArchiveBuilderBuilder {
 fn create_mingw_dll_import_lib(
     sess: &Session,
     lib_name: &str,
-    import_name_and_ordinal_vector: Vec<(String, Option<u16>)>,
+    items: Vec<ImportLibraryItem>,
     output_path: &Path,
 ) {
     let def_file_path = output_path.with_extension("def");
 
     let def_file_content = format!(
         "EXPORTS\n{}",
-        import_name_and_ordinal_vector
+        items
             .into_iter()
-            .map(|(name, ordinal)| {
+            .map(|ImportLibraryItem { name, ordinal, .. }| {
                 match ordinal {
                     Some(n) => format!("{name} @{n} NONAME"),
                     None => name,

--- a/compiler/rustc_codegen_ssa/src/back/link.rs
+++ b/compiler/rustc_codegen_ssa/src/back/link.rs
@@ -45,7 +45,7 @@ use rustc_target::spec::{
 use tempfile::Builder as TempFileBuilder;
 use tracing::{debug, info, warn};
 
-use super::archive::{ArchiveBuilder, ArchiveBuilderBuilder};
+use super::archive::{ArchiveBuilder, ArchiveBuilderBuilder, ImportLibraryItem};
 use super::command::Command;
 use super::linker::{self, Linker};
 use super::metadata::{create_wrapper_file, MetadataPosition};
@@ -498,16 +498,35 @@ fn create_dll_import_libs<'a>(
 
             let mingw_gnu_toolchain = common::is_mingw_gnu_toolchain(&sess.target);
 
-            let import_name_and_ordinal_vector: Vec<(String, Option<u16>)> = raw_dylib_imports
+            let items: Vec<ImportLibraryItem> = raw_dylib_imports
                 .iter()
                 .map(|import: &DllImport| {
                     if sess.target.arch == "x86" {
-                        (
-                            common::i686_decorated_name(import, mingw_gnu_toolchain, false),
-                            import.ordinal(),
-                        )
+                        ImportLibraryItem {
+                            name: common::i686_decorated_name(
+                                import,
+                                mingw_gnu_toolchain,
+                                false,
+                                false,
+                            ),
+                            ordinal: import.ordinal(),
+                            symbol_name: import.is_missing_decorations().then(|| {
+                                common::i686_decorated_name(
+                                    import,
+                                    mingw_gnu_toolchain,
+                                    false,
+                                    true,
+                                )
+                            }),
+                            is_data: !import.is_fn,
+                        }
                     } else {
-                        (import.name.to_string(), import.ordinal())
+                        ImportLibraryItem {
+                            name: import.name.to_string(),
+                            ordinal: import.ordinal(),
+                            symbol_name: None,
+                            is_data: !import.is_fn,
+                        }
                     }
                 })
                 .collect();
@@ -515,7 +534,7 @@ fn create_dll_import_libs<'a>(
             archive_builder_builder.create_dll_import_lib(
                 sess,
                 &raw_dylib_name,
-                import_name_and_ordinal_vector,
+                items,
                 &output_path,
             );
 

--- a/compiler/rustc_codegen_ssa/src/common.rs
+++ b/compiler/rustc_codegen_ssa/src/common.rs
@@ -187,12 +187,15 @@ pub fn i686_decorated_name(
     dll_import: &DllImport,
     mingw: bool,
     disable_name_mangling: bool,
+    force_fully_decorated: bool,
 ) -> String {
     let name = dll_import.name.as_str();
 
-    let (add_prefix, add_suffix) = match dll_import.import_name_type {
-        Some(PeImportNameType::NoPrefix) => (false, true),
-        Some(PeImportNameType::Undecorated) => (false, false),
+    let (add_prefix, add_suffix) = match (force_fully_decorated, dll_import.import_name_type) {
+        // No prefix is a bit weird, in that LLVM/ar_archive_writer won't emit it, so we will
+        // ignore `force_fully_decorated` and always partially decorate it.
+        (_, Some(PeImportNameType::NoPrefix)) => (false, true),
+        (false, Some(PeImportNameType::Undecorated)) => (false, false),
         _ => (true, true),
     };
 

--- a/compiler/rustc_session/src/cstore.rs
+++ b/compiler/rustc_session/src/cstore.rs
@@ -130,6 +130,11 @@ impl DllImport {
             None
         }
     }
+
+    pub fn is_missing_decorations(&self) -> bool {
+        self.import_name_type == Some(PeImportNameType::Undecorated)
+            || self.import_name_type == Some(PeImportNameType::NoPrefix)
+    }
 }
 
 /// Calling convention for a function defined in an external library.

